### PR TITLE
[ AFTER: https://github.com/ceph/ceph/pull/32441 ] nautilus: rgw/kafka: add kafka endpoint support

### PR DESCRIPTION
**Pull request by yuvalif**
TODO:
- [x] wait for https://github.com/ceph/ceph/pull/31834 to be merged into master, so it could be backported to nautilus as part of this PR
- [x] add to release notes the list of platforms on which kafka is not supported out of the box and will require more updated librdkafka
backport trackers:

* https://tracker.ceph.com/issues/43156 (for the 100% CPU issue)
* the actual kafka work has no tracker, was originally done in: https://github.com/ceph/ceph/pull/30960

---

* adds kafka endpoint support to pubsub/notifications
* include the fix to: https://tracker.ceph.com/issues/43104
* information on how to use kafak endpoints is in the docs

> kafka security will be backported once https://github.com/ceph/ceph/pull/31834 is merged

### CODE REVIEW

**Comment by yuvalif** [comment]
@tchaikov kafka does not build on ubuntu xenial, which block any kafka PR (jenkins is running on xenial). For master I modified "src/script/run-make.sh", but in nautilus this file does not exist. any equivalent file that can be used?

**Comment by tchaikov** [comment]
@yuvalif you might want to update `run-make-check.sh` accordingly.

**Comment by yuvalif** [comment]
> @yuvalif you might want to update `run-make-check.sh` accordingly. but if kafka does not build on ubuntu xenial, why not fix this as suggested by @cbodley at https://github.com/ceph/ceph/pull/30960/files#r349269968 ?

what @cbodley suggested is that we should fail the build during cmake with the appropriate error message (=invalid librdkafka version), instead of during compilation. but this won't fix the issue that jenking is failing.
will do that in addition.

**Comment by yuvalif** [comment]
> > @yuvalif you might want to update `run-make-check.sh` accordingly. but if kafka does not build on ubuntu xenial, why not fix this as suggested by @cbodley at https://github.com/ceph/ceph/pull/30960/files#r349269968 ?
> 
> what @cbodley suggested is that we should fail the build during cmake with the appropriate error message (=invalid librdkafka version), instead of during compilation. but this won't fix the issue that jenking is failing.
> will do that in addition.

done here: https://github.com/ceph/ceph/pull/32190

**Comment by smithfarm** [comment]
@yuvalif Did all the commits apply without conflicts? I tried to cherry-pick 354d775 to nautilus and got a conflict.

Please see https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst#cherry-picking-rules

**Comment by smithfarm** [comment]
@yuvalif Please

* include conflict resolution sections on all cherry-picks that needed manual adjustment
* include all relevant backport tracker URL in the PR description
* update all the backport tracker issues - see how I just did it at https://tracker.ceph.com/issues/43156
* parent PR URL and parent tracker URL are nice to have, but optional

Thanks for backporting.

**Comment by smithfarm** [comment]
In the past, when "fresh" features have been backported to stable releases, this has led to an "onslaught" of many backport PRs in an effort to keep up with the development of the feature in master.

Would it make sense to let this feature "bake" in master for some time before backporting it to nautilus?

**Comment by yuvalif** [comment]
> @yuvalif Did all the commits apply without conflicts? I tried to cherry-pick 354d775 to nautilus and got a conflict.
> 
> Please see https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst#cherry-picking-rules

@smithfarm I fixed the conflicts manually as part of the cherry-pick

**Comment by smithfarm** [comment]
> I fixed the conflicts manually as part of the cherry-pick

Probably the cherry-picking rules in SubmittingPatches-backports.rst need to be clarified. This section covers it (just ignore the mentions of the script if you prefer not to use it):

https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst#conflict-resolution

**Comment by yuvalif** [comment]
> In the past, when "fresh" features have been backported to stable releases, this has led to an "onslaught" of many backport PRs in an effort to keep up with the development of the feature in master.
> 
> Unfortunately, this does not work well with the existing backporting workflows - all the commits related to the feature have to be merged to "nautilus" in the same order that they get merged to "master", but we don't have any "order-enforcing" mechanism in the backport tracker. So, inevitably, the backports start to get merged out of order...
> 
> Would it make sense to let this feature "bake" in master for some time before backporting it to nautilus?

currently, the only other PR that should have been backported together with this one is: https://github.com/ceph/ceph/pull/31834 
but it is still not merged to master.
I would like to avoid too much delays with backporting of kafka, but don't mind waiting little longer (hopefully the "kafka security" PR will be merged soon...)

**Comment by smithfarm** [comment]
@yuvalif We could keep this PR open and cherry-pick kafka-related fixes into it as they go into master. Once you think the feature is stable, it could be merged.

**Comment by yuvalif** [comment]
> @yuvalif Please
> 
> * include conflict resolution sections on all cherry-picks that needed manual adjustment

will do

> * include all relevant backport tracker URL in the PR description

there is only one tracker issue (the 2nd commit which is bug fix on the 1st commit)
the 1st commit does not have any tracker - since it is a new feature

> * update all the backport tracker issues - see how I just did it at https://tracker.ceph.com/issues/43156

will do

> * parent PR URL and parent tracker URL are nice to have, but optional
> 
> Thanks for backporting.

**Comment by yuvalif** [comment]
> @yuvalif We could keep this PR open and cherry-pick kafka-related fixes into it as they go into master. Once you think the feature is stable, it could be merged if the dev leads agree.

the remaining kafka work on security was merged to master and cherry-picked here.

**Comment by yuvalif** [comment]
> Please add the appropriate conflict resolution message to the first commit:
> 
> ```
> Conflicts:
>         ceph.spec.in
>         debian/control
>         do_cmake.sh
>         doc/radosgw/notifications.rst
>         src/test/rgw/rgw_multi/tests_ps.py
> ```

done

**Comment by yuvalif** [comment]
jenkins test make check

File: PendingReleaseNotes [hunk 1]
```diff
@@ -131,3 +131,7 @@
   The above options can be set with::
 
     ceph config set global <option> <value>
+
+* Bucket notifications now support Kafka endpoints. This require librdkafka of version 0.9.2 and up.
+  Note that Ubuntu 16.04.6 LTS (Xenial Xerus) has an older version of libradkafka, and would
```
Comment (smithfarm) [inline PendingReleaseNotes]: is `libradkafka` a typo? (The previous line has `librdkafka`.)

Comment (yuvalif) [reply PendingReleaseNotes]: yes, should be "librdkafka"

Comment (yuvalif) [reply PendingReleaseNotes]: fixed

File: PendingReleaseNotes [hunk 2]
```diff
@@ -138,3 +138,7 @@
 * The MGR now accepts 'profile rbd' and 'profile rbd-read-only' user caps.
   These caps can be used to provide users access to MGR-based RBD functionality
   such as 'rbd perf image iostat' an 'rbd perf image iotop'.
+
+* Bucket notifications now support Kafka endpoints. This require librdkafka of version 0.9.2 and up.
+  Note that Ubuntu 16.04.6 LTS (Xenial Xerus) has an older version of librdkafka, and would
+  require an update to the library.
```
Comment (smithfarm) [inline PendingReleaseNotes]: the release note was added under the heading `12.2.7` which is wrong. Please put it under a new heading (at the top of the file) - the new heading should be:

```
14.2.7
------
```

**Comment by yuvalif** [comment]
this PR deviated too much from nautilus. will close in favor of: https://github.com/ceph/ceph/pull/32960

**Pull request closed without merging**